### PR TITLE
[CALCITE-6296] Support IS NULL in Arrow adapter

### PR DESCRIPTION
**Pull request by timgrein**
Add the unary `isnull` Gandiva operator to `ArrowTranslator`.

### CODE REVIEW

**Review by asolimando — APPROVED** [review]
LGTM

File: arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java
Comment (caicancai) [inline arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:332]: Just a question, why not add a new test

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java
Comment (caicancai) [inline arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:192]: Your javadoc format should be consistent

File: arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java
Comment (asolimando) [reply arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:332]: That test was already covering exactly this missing feature, and was meant to be fixed upon implementing it.
I think it's OK unless you see the need of a new test around `IS NULL` which is not covered here.

Comment (caicancai) [reply arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:332]: Because this test involves two jira cases, and both are bug fixes, I think it should be tested separately and add jira links, which may help subsequent developers understand

Comment (asolimando) [reply arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:332]: When you say it involves two Jira cases, you mean that it was added in one past ticket and that we are modifying it in the context of this one?

The ticket has been originally inserted as disabled (per my request), I have subsequently enabled it when the fallback on `Enumerable` was added, with the guard on the present ticket exactly to have the test fixed as in this PR.

The value of the previous version of test was to exercise the fallback while the feature wasn't supported, but now that part is not needed anymore and it can be removed.

For clarity we can add a link to the present Jira ticket as it fully and exclusively covers the feature, and the old tickets don't apply anymore.

Would that be OK for you, @caicancai?

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java
Comment (asolimando) [reply arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:192]: The comment applies also to the other functions that were introduced

Comment (timgrein) [reply arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java]: Adjusted with [Adjust comment to follow the format of the other comments](https://github.com/apache/calcite/pull/3858/commits/3c48ffe1b54e1c2e1db5ff2a40089638bd67106a). I guess you meant putting it on one line?

**Comment by timgrein** [comment]
> Edit: there seems to a problem with Sonar in CI which doesn't look like a transient failure, can you take a look @timgrein? Did you experience any issues locally?

I couldn't reproduce it locally. I ran the following commands:

`./gradlew clean ':arrow:test' --no-build-cache`

and

`./gradlew build`. Any other commands you want to share, which might be helpful to investigate this? @asolimando 

Both succeeded without the issue from CI, I'll check, if it appears again.

**Comment by caicancai** [comment]
> > Edit: there seems to a problem with Sonar in CI which doesn't look like a transient failure, can you take a look @timgrein? Did you experience any issues locally?
> 
> I couldn't reproduce it locally. I ran the following commands:
> 
> `./gradlew clean ':arrow:test' --no-build-cache`
> 
> and
> 
> `./gradlew build`. Any other commands you want to share, which might be helpful to investigate this? @asolimando
> 
> Both succeeded without the issue from CI, I'll check, if it appears again.

2024-07-16 23-36-14屏幕截图

You can look at the cli log.

**Comment by timgrein** [comment]
> > > Edit: there seems to a problem with Sonar in CI which doesn't look like a transient failure, can you take a look @timgrein? Did you experience any issues locally?
> > 
> > 
> > I couldn't reproduce it locally. I ran the following commands:
> > `./gradlew clean ':arrow:test' --no-build-cache`
> > and
> > `./gradlew build`. Any other commands you want to share, which might be helpful to investigate this? @asolimando
> > Both succeeded without the issue from CI, I'll check, if it appears again.
> 
> 2024-07-16 23-36-14屏幕截图
> 
> You can look at the cli log. Integration/jenkins/pr-head may occasionally fail. If the log exception is not very strange, I think this will not affect the merge.

Thanks for the context! Seems like it succeeded now.

File: arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java
Comment (NobiGo) [inline arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:329]: Maybe we can add another test for "a is null and b is null" or "a is null or b=c", To cover when the IS NULL is in a complicated condition.

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java
Comment (asolimando) [inline arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:133]: I realize now that "binary" is not correct, it can be unary too, can we amend as `"Unsupported operator " + node`?

File: arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java
Comment (timgrein) [reply arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:329]: Good point! Added tests with [Add Conjunctive IsNull Filter tests](https://github.com/apache/calcite/pull/3858/commits/6d31f70295e36529ff5691ae1d52f1a4c95389c8)

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java
Comment (asolimando) [inline arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:133]: I realize now that "binary" is not correct, it can be unary too, can we amend it as follows?
```suggestion
      throw new UnsupportedOperationException("Unsupported operator " + node);
```

Comment (timgrein) [reply arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java]: Good catch, adjusted with [Adjust UnsupportedOperationException message](https://github.com/apache/calcite/pull/3858/commits/96ddd0ce69fe9f70f96862d4bdc77a304b02aab5) (saw your old comment first :) )

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTable.java
Comment (asolimando) [inline arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTable.java:132]: This is not necessarily an equality, right?

File: arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java
Comment (asolimando) [inline arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:328]: We generally use this form to add Jira links for test cases (you can grep the codebase to see, if needed):
```suggestion
  /** Test case for
  * <a href="https://issues.apache.org/jira/browse/CALCITE-6296">[CALCITE-6296]
  * Support IS NULL in Arrow adapter</a>. */
```

File: arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java
Comment (asolimando) [reply arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:333]: @timgrein please check https://github.com/apache/calcite/pull/3858/files#r1680614476 for the comment format

File: arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java
Comment (asolimando) [inline arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:345]: Let's add the jira link here too, for uniformity

**Review by asolimando — CHANGES_REQUESTED** [review]
Few minor changes emerged during review, let's address them, but they are minor and we are converging

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTable.java
Comment (timgrein) [reply arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTable.java]: Nope, it isn't. Gandiva refers to it as a function, but I think `operator` definitely makes way more sense. Adjusted with [Rename equality to operator in ArrowTable](https://github.com/apache/calcite/pull/3858/commits/a7c195b43cc0e7b1d6c84bcdd1873efebb3be0fb)

File: arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java
Comment (timgrein) [reply arrow/src/test/java/org/apache/calcite/adapter/arrow/ArrowAdapterTest.java:350]: [Add test case comment to testConjunctiveIsNullFilters](https://github.com/apache/calcite/pull/3858/commits/634467a9e6bf98db8d4cf07f9fab19edf86eec0c)

**Comment by timgrein** [comment]
> Few minor changes emerged during review, let's address them, but they are minor and we are converging

Thanks for the detailed review @asolimando, already helped a lot! Addressed your changes, let me know, if there's something else.

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java
Comment (caicancai) [inline arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:205]: Are you sure this conversion will be successful? Do you need to make a judgment?

**Comment by caicancai** [comment]
Mostly looks good, just a few minor comments

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java
Comment (timgrein) [reply arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:205]: I checked other places, where the same cast happens:
- [Cast in DruidExpressions](https://github.com/apache/calcite/blob/7d9b639cd41ff7b3e3c81a705da5a38308c0b8ac/druid/src/main/java/org/apache/calcite/adapter/druid/DruidExpressions.java#L109-L110)
- [Cast in DruidQuery](https://github.com/apache/calcite/blob/7d9b639cd41ff7b3e3c81a705da5a38308c0b8ac/druid/src/main/java/org/apache/calcite/adapter/druid/DruidQuery.java#L403-L404)
- [Cast in GeodeFilter](https://github.com/apache/calcite/blob/7d9b639cd41ff7b3e3c81a705da5a38308c0b8ac/geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeFilter.java#L372-L374)
- [Casts in ProjectAggregateMergeRule](https://github.com/apache/calcite/blob/7d9b639cd41ff7b3e3c81a705da5a38308c0b8ac/core/src/main/java/org/apache/calcite/rel/rules/ProjectAggregateMergeRule.java#L95-L106)

I can add an `instanceof` check, but this might be unnecessary? At least that's the impression I get from the examples above.

Comment (asolimando) [reply arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:205]: I don't think it's necessary, as you have noticed it's a common pattern to rely on `SqlKind` to infer the type of an object instead of `instanceof`, for me it's good as-is

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java
Comment (asolimando) [inline arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:214]: Nit: I just noticed this is used in a single place, I guess we can inline as it's a single line and the size would be almost the same.

**Review by asolimando — APPROVED** [review]
LGTM, one final nitpick but it's optional, all other comments seem to have been addressed.

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java
Comment (timgrein) [reply arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:214]: I think we'll need at least one more unary operator afterwards, which I'll also look into: [Support IS NOT NULL in Arrow adapter](https://issues.apache.org/jira/browse/CALCITE-6295?jql=project%20%3D%20CALCITE%20AND%20component%20%3D%20arrow-adapter), so I guess we can keep it? WDYT?

**Review by caicancai — APPROVED** [review]
LGTM

File: arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java
Comment (caicancai) [reply arrow/src/main/java/org/apache/calcite/adapter/arrow/ArrowTranslator.java:205]: @asolimando thank you for the explanation

**Comment by asolimando** [comment]
If @NobiGo is fine as well with the current status of the PR, I will ask you @timgrein to squash everything into a single commit and I will then merge. Please wait for the final confirmation before squashing as it makes incremental reviewing way harder, thanks!

**Review by NobiGo — APPROVED** [review]

**Comment by asolimando** [comment]
@timgrein, can you please squash the commits into a single one, having "[CALCITE-6296] Support IS NULL in Arrow adapter" as commit message? After that I will merge, thanks!

**Pull request merged**
